src/main/java/org/broadinstitute/hellbender/tools/walkers/coverage/CoverageUtils.java

```diff
@@ -0,0 +1,317 @@
+                    }
+                }
+                break;
+
+            // TODO see reconcile FragmentUtils.create() and its various idiosyncrasies to reenable this feature
+            case COUNT_FRAGMENTS: // ignore base identities and put in FIRST base that passes filters:
+                throw new GATKException("Fragment based counting is currently unsupported");
+
+            case COUNT_FRAGMENTS_REQUIRE_SAME_BASE:
+                throw new GATKException("Fragment based counting is currently unsupported");
+
+            default:
+                throw new UserException("Must use valid CountPileupType");
+        }
+
+        for (PileupElement e : countPileup) {
+            SAMReadGroupRecord readGroup = ReadUtils.getSAMReadGroupRecord(e.getRead(), header);
+            Utils.nonNull(readGroup, String.format("Read %s was missing read group information", e.getRead()));
+
+            // uniqueReadGroupID is unique across the library, read group ID, and the sample
+            String uniqueReadGroupId = readGroup.getSample() + "_" + readGroup.getReadGroupId() + "_" + readGroup.getLibrary() + "_" + readGroup.getPlatformUnit();
+            int[] counts = countsByRGName.get(uniqueReadGroupId);
+            if (counts == null) {
+                counts = new int[6];
+                countsByRGName.put(uniqueReadGroupId, counts);
+                RGByName.put(uniqueReadGroupId, readGroup);
+            }
+
+            updateCounts(counts, e);
+        }
+
+        for (String readGroupId : RGByName.keySet()) {
+            countsByRG.put(RGByName.get(readGroupId), countsByRGName.get(readGroupId));
+        }
+
+        return countsByRG;
+    }
+
+    // Applies the provided mapping and base quality filters to the provided read
+    private static boolean filterElementFromCount(final PileupElement e, final byte minBaseQ, final byte maxBaseQ) {
+        return ( e.getQual() >= minBaseQ && e.getQual() <= maxBaseQ || e.isDeletion() );
+    }
+
+    private static void updateCounts(int[] counts, PileupElement e) {
+        if ( e.isDeletion() ) {
+            counts[BaseUtils.Base.D.ordinal()]++;
+        } else if ( BaseUtils.basesAreEqual(BaseUtils.Base.N.base, e.getBase()) ) {
+            counts[BaseUtils.Base.N.ordinal()]++;
+        } else {
+            try {
+                counts[BaseUtils.simpleBaseToBaseIndex(e.getBase())]++;
+            } catch (ArrayIndexOutOfBoundsException exc) {
+                throw new GATKException("Expected a simple base, but actually received"+(char)e.getBase());
+            }
+        }
+    }
+
+    /**
+     * This method is used to construct the bins we use to compute coverage histograms.
+     *
+     * Returns the coverage "leftEndpoints" which correspond to the lowest value inclusive to be included in the given
+     * histogram bins. This method will attempt to distribute the bins between the specified values based a logarithmic
+     * distribution?
```

Comment:
Because some drunk and bored scribe copying the documentation late into the night from their monastery decided to write their own commentary in the margins and it is now accepted as part of the document itself....